src/Init/Internal/ExtrinsicTermination.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Reichert
+-/
+module
+
+prelude
+import Init.Classical
+public import Init.NotationExtra
+
+set_option doc.verso true
+
+/-!
+# Fixpoints with extrinsic proofs of termination
+
+This module provides a fixpoint combinator that does not require an immediate *intrinsic*
+termination proof, but lends itself to formal verification given a deferred *extrinsic* termination
+proof.
+
+This fixpoint combinator can be seen as a compromise between using a `partial` function, which is
+fully opaque, and using well-founded recursion, which requires an intrinsic termination proof.
+
+The soundness (in a wide sense, including compatibility of kernel and compiler semantics) relies on
+the same assumption as the one WF preprocessing is built upon: When two fixpoint-defining
+functionals are propositionally equal, then the observable runtime behavior of their fixpoints is
+identical.
+-/
+
+variable {α : Sort _} {β : α → Sort _} {C : α → Sort _} {C₂ : (a : α) → β a → Sort _}
+
+namespace Std.Internal
+
+@[specialize]
+public partial def opaqueFix [∀ x, Nonempty (C x)] (F : (x : α) → ((y : α) → C y) → C x) (x : α) : C x :=
+  F x (opaqueFix F)
+
+@[expose]
+public def TerminatesTotally (F : (x : α) → ((y : α) → C y) → C x) : Prop :=
+  ∃ (r : α → α → Prop) (F' : (x : α) → ((y : α) → r y x → C y) → C x), WellFounded r ∧ ∀ x G, F x G = F' x (fun x _ => G x)
+
+/-
+SAFE assuming that the code generated by iteration over `F` is compatible
+with the kernel semantics of iteration over `F' : (x : α) → ((y : α) → r y x → C y) → C x`
+for all `F'` as in `TerminatesTotally`.
+-/
+@[implemented_by opaqueFix]
+public def extrinsicFix [∀ x, Nonempty (C x)] (F : (x : α) → ((y : α) → C y) → C x) (x : α) : C x :=
+  open scoped Classical in
+  if h : TerminatesTotally F then
+    let F' := h.choose_spec.choose
+    let h := h.choose_spec.choose_spec
+    h.1.fix F' x
+  else
+    opaqueFix F x
```

Comment:
Why not `Classical.ofNonempty`? That's what `opaqueFix F x` will look like in the logic. `#print axioms opaqueFix` even reports `Classical.choice`.

At runtime, we will never see `extrinsicFix` (it is noncomputable anyway), so there is no point about writing `opaqueFix` here other than to let reader scratch their head.